### PR TITLE
Fix assign for components of vector function spaces

### PR DESCRIPTION
**Pull request by connorjward**
@JHopeCollins this should hopefully fix things for you.

### CODE REVIEW

**Comment by connorjward** [comment]
> Thanks for getting on this so quick Connor. In ` test_mixed_expressions_indexed_fs` could there also be a test for assigning to a scalar function from a component of a mixed function, as well as the tests for assigning to a component of a mixed function from a scalar function please?

Sure. Done.

File: firedrake/assign.py [hunk 1]
```diff
@@ -131,11 +131,11 @@ def __init__(self, assignee, expression, subset=None):
             expression = expression.function
         expression = as_ufl(expression)
 
-        if not all(c.function_space() == assignee.function_space()
+        if not all(c.ufl_element() == assignee.ufl_element()
```
Comment (JDBetteridge) [inline firedrake/assign.py:134]: I still have concerns about users trying to add functions on different meshes, but agree this shouldn't prevent this fix going in. We should open an issue with an MFE.

File: tests/regression/test_expressions.py [hunk 2]
```diff
@@ -169,10 +169,60 @@ def test_tensor_expressions(expr, tfunctions):
...
+    f.sub(1).assign(two)
+    assert evaluate(f.dat.data, (0, 2))
+    f.sub(0).assign(one)
+    assert evaluate(f.dat.data, (1, 2))
+
+    one.assign(2*f.sub(0) + 1)
+    assert evaluate(one.dat.data, 3)
+
+    two += f.sub(1)
+    assert evaluate(two.dat.data, 4)
```
Comment (JDBetteridge) [inline tests/regression/test_expressions.py:226]: I think @JHopeCollins might have some more tests to contribute, but these can be added at a later date

**Review by JDBetteridge — APPROVED** [review]

File: firedrake/assign.py (hunk 1 above)
Comment (connorjward) [reply firedrake/assign.py]: I've pushed a commit that does this.

**Review by JHopeCollins — APPROVED** [review]
Yes, happy with this

**Pull request merged**